packages/protocol/contracts/layer1/shasta/impl/Inbox.sol

```diff
@@ -0,0 +1,725 @@
+            ExceedsUnfinalizedProposalCapacity()
+        );
+    }
+
+    /// @dev Processes forced inclusion if required
+    /// @param _config The configuration parameters.
+    /// @param _coreState The core state.
+    /// @return The updated core state.
+    function _processForcedInclusion(
+        Config memory _config,
+        CoreState memory _coreState
+    )
+        private
+        returns (CoreState memory)
+    {
+        IForcedInclusionStore store = IForcedInclusionStore(_config.forcedInclusionStore);
+
+        if (!store.isOldestForcedInclusionDue()) {
+            return _coreState;
+        }
+
+        IForcedInclusionStore.ForcedInclusion memory forcedInclusion =
+            store.consumeOldestForcedInclusion(msg.sender);
+
+        return _propose(_config, _coreState, forcedInclusion.blobSlice, true);
+    }
+
+    /// @dev Validates the inputs for prove function
+    /// @param _proposals The proposals to prove.
+    /// @param _claims The claims for the proposals.
+    function _validateProveInputs(
+        Proposal[] memory _proposals,
+        Claim[] memory _claims
+    )
+        private
+        pure
+    {
+        require(_proposals.length == _claims.length, InconsistentParams());
+        require(_proposals.length != 0, EmptyProposals());
+    }
+
+    /// @dev Stores claim records and emits events
+    /// @param _config The configuration parameters.
+    /// @param _claimRecords The claim records to store.
+    function _storeClaimRecords(
+        Config memory _config,
+        ClaimRecord[] memory _claimRecords
+    )
+        private
+    {
+        for (uint256 i; i < _claimRecords.length; ++i) {
+            bytes32 claimRecordHash = keccak256(abi.encode(_claimRecords[i]));
+
+            ProposalRecord storage proposalRecord =
+                _proposalRecord(_config, _claimRecords[i].claim.proposalId);
+            proposalRecord.claimHashLookup[_claimRecords[i].claim.parentClaimHash].claimRecordHash =
+                claimRecordHash;
+
+            emit Proved(_claimRecords[i]);
+        }
+    }
+
+    /// @dev Verifies the proposal is the last one proposed
+    /// @param _config The configuration parameters.
+    /// @param _proposals The proposals array to verify.
+    function _verifyLastProposal(
+        Config memory _config,
+        Proposal[] memory _proposals
+    )
+        private
+        view
+    {
+        bytes32 storedNextProposalHash = _proposalRecord(_config, _proposals[0].id + 1).proposalHash;
+
+        if (storedNextProposalHash == bytes32(0)) {
+            // Next slot is empty, only one proposal expected
+            require(_proposals.length == 1, IncorrectProposalCount());
+        } else {
+            // Next slot is occupied, need to prove it contains a smaller proposal id
+            require(_proposals.length == 2, IncorrectProposalCount());
+            require(_proposals[1].id < _proposals[0].id, NextProposalIdSmallerThanLastProposalId());
```

Comment:
A little bit confused here, if `_proposals[1].id < _proposals[0].id`, how can `_checkProposalHash(_config, _proposals[1]);` promise that the `_proposals` are the last ones proposed? Shall it be:

```solidity
require(_proposals[1].id > _proposals[0].id, NextProposalIdSmallerThanLastProposalId());
_verifyLastProposal(_config, _proposals[1]);
```